### DISCUSSION TITLE
Bun provider

### DISCUSSION THREAD

**Discussion by Milo123459** [railwayapp/nixpacks#261] [Ideas]
Just posting in its own discussion as it is pretty lengthy. 

- How do we know when to use Bun?
- Should we even use Bun?

Bun: https://bun.sh

Bun currently works as a drop-in replacement for Node and is significantly faster.

**Comment by coffee-cup** [comment]
The team has discussed this internally a bit.

There is a PR to add Bun to nixpkgs https://github.com/NixOS/nixpkgs/pull/180299 so can pull from there once it is merged.

Bun is still in beta so I don't think we should be making it the default for the Node provider anytime soon. But it would be great if there was a Bun provider that would install and run with it.

For detection I think we can look for `bunfig.toml` or `bun.lockb` files. For reference this is what the filesystem looks like after running `bun create next`

![image](https://user-images.githubusercontent.com/3044853/177836782-bf38935b-840d-4985-9f31-4a8698e5d402.png)

**Reply by Milo123459** [reply]
I agree with holding off on making it the default. This *could* be something we do later down the line. I guess with the lockfiles being generated, we could just make another package manager in the node provider. 

PS: When we say make default, do we mean get rid of support for other package managers?

**Reply by nebulatgs** [reply]
Probably meant replacing `npm` as the default

**Reply by Milo123459** [reply]
That's what I was thinking. Just wanted to check though.

**Reply by coffee-cup** [reply]
Yeah we can add it to the NodeProvider just by looking at the lockfile.

Hopefully it gets merged into nixpkgs soon

**Reply by Milo123459** [reply]
I'll subscribe to the PR. When it gets merged, I'll add support for it but won't make it the default.

**Comment by Milo123459** [comment]
This has now been implemented.
